File: glv/commit.py
#
# Copyright (c) 2021 Bahtiar `kalkin-` Gadimov.
#
# This file is part of Git Log Viewer
# (see https://github.com/kalkin/git-log-viewer).
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.
#
''' New core api objects '''

import logging
from collections import namedtuple
from datetime import datetime, timezone
from typing import Optional

import babel
import git

from glv import vcs

LOG = logging.getLogger('glv')

Commit = namedtuple(
    'Commit',
    [
        'oid',
        'short_id',
        'author_name',
        'author_email',
        'author_date',
        'author_rel_date',
        'committer_name',
        'committer_email',
        'committer_date',
        'subject',
        'type_icon',
        # optional
        'above',
        'bellow',
        'branches',
        'children',
        'has_level',
        'is_commit_link',
        'is_fork_point',
        'is_head',
        'is_merge',
        'level',
        'references',
        'tags',
    ],
    defaults=[
        None,  # above
        None,  # bellow
        [],  # branches
        [],  # children
        False,  # has_level
        False,  # is_commit_link
        False,  # is_fork_point
        False,  # is_head
        False,  # is_merge
        0,  # level
        [],  # references
        [],  # tags
    ])

REV_FORMAT = '%h%x00'  # abbrev hash
REV_FORMAT += '%P%x00'  # parents
REV_FORMAT += '%D%x00'  # references
REV_FORMAT += '%aN%x00%aE%x00%aI%x00'  # mailmap author name/mail/date
REV_FORMAT += '%cN%x00%cE%x00%cI%x00'  # mailmap committer name/mail/date
REV_FORMAT += '%s%x00'  # subject


def to_commit(working_dir: str, oid: str, **kwargs) -> Commit:
    ''' Return data for one specific commit '''
    git_cmd = git.cmd.Git(working_dir=working_dir)
    git_p = git_cmd.rev_list(oid,
                             format=REV_FORMAT,
                             max_count=1,
                             as_process=True)
    stream = git_p.proc.stdout
    line1 = stream.readline()
    line2 = stream.readline()
    return parse_commit(working_dir, line1, line2, **kwargs)


def parse_commit(working_dir: str,
                 line1: bytes,
                 data_line: bytes,
                 above_commit: Commit = None,
                 level=-1,
                 is_commit_link=False) -> Commit:
    ''' Just a helpful wrapper.  '''

    # pylint: disable=too-many-locals,too-many-arguments
    assert line1.startswith(b'commit ')  # nosec
    oid = line1.partition(b' ')[2].decode().strip()
    short_id, parents_record, reference_record, \
            auth_name, auth_email, auth_date, \
            com_name, com_email, com_date, \
            subject, _ = [
            x.decode() for x in data_line.split(b'\0')
        ]

    auth_rel_date = _to_rel_date(datetime.fromisoformat(auth_date))
    is_head = False
    references = []
    branches = []
    tags = []
    for smth in reference_record.split(', '):
        if smth == 'HEAD':
            is_head = True
        elif smth.startswith('HEAD -> '):
            is_head = True
            _, _, branch = smth.partition(' -> ')
            references.append(branch)
            branches.append(branch)
        elif smth.startswith('tag: '):
            _, _, tag = smth.partition(': ')
            references.append(tag)
            tags.append(tag)
        else:
            references.append(smth)
            branches.append(smth)

    _children: list[str] = []
    parents = parents_record.split(' ')
    bellow = None
    children = []
    is_merge = False
    above = None
    if above_commit:
        above = above_commit.oid

    git_cmd = git.cmd.Git(working_dir)
    is_fork_point = False
    if above_commit and above_commit.children and above_commit.level == level:
        try:
            git_cmd.merge_base(oid, above_commit.children[0], is_ancestor=True)
            is_fork_point = True
        except git.GitCommandError:
            pass

    if parents:
        bellow = parents[0]
        if len(parents) > 1:
            children = parents[1:]
            is_merge = True

    type_icon_level = ''
    if level > 0:
        type_icon_level = level * '│ '
    type_icon = type_icon_level + _type_icon(bellow, is_commit_link)

    return Commit(oid,
                  short_id,
                  auth_name,
                  auth_email,
                  auth_date,
                  auth_rel_date,
                  com_name,
                  com_email,
                  com_date,
                  subject,
                  type_icon,
                  above=above,
                  bellow=bellow,
                  level=level,
                  has_level=level >= 0,
                  is_commit_link=is_commit_link,
                  is_fork_point=is_fork_point,
                  branches=branches,
                  children=children,
                  is_head=is_head,
                  is_merge=is_merge,
                  references=references,
                  tags=tags)


def commits_for_range(  # pylint: disable=too-many-arguments
        working_dir: str,
        rev_range: str,
        above_commit: str = None,
        level: int = 0,
        paths: list[str] = None,
        rev_list_args: dict = None) -> list[Commit]:
    ''' Return commits for a range '''
    # pylint: disable=too-many-locals

    rev_list_args = rev_list_args or {}
    git_cmd = git.cmd.Git(working_dir=working_dir)
    args = [rev_range]
    if paths:
        args.append('--')
        args += paths

    cmd_proc = git_cmd.rev_list(args,
                                first_parent=True,
                                format=REV_FORMAT,
                                **rev_list_args,
                                as_process=True)
    result = []
    parse_data = False
    stream = cmd_proc.proc.stdout
    above = above_commit
    while True:
        assert not parse_data  # nosec
        line1 = stream.readline()
        if not line1:
            break
        line2 = stream.readline()
        if not line2:
            break
        _commit = parse_commit(working_dir,
                               line1,
                               line2,
                               above_commit=above,
                               level=level)
        result.append(_commit)
        above = _commit

    return result


def merge_base(working_dir: str, *oids) -> Optional[str]:
    ''' Return the mergebase commit id '''
    git_cmd = git.cmd.Git(working_dir=working_dir)
    try:
        return git_cmd.merge_base(*oids).strip()
    except git.GitCommandError:
        return None


def child_history(working_dir: str, commit: Commit) -> list[Commit]:
    ''' Return history for the fist child commit '''
    assert commit.is_merge  # nosec
    bellow: str = commit.bellow
    first_child: str = commit.children[0]
    end: str = merge_base(working_dir, bellow, first_child)
    revision = '%s..%s' % (end, first_child)
    if not end:
        revision = '%s' % first_child
    LOG.debug('commits_for_range(%r)', revision)
    result = commits_for_range(working_dir,
                               revision,
                               above_commit=commit,
                               level=commit.level + 1)

    end_commit = result[-1]
    if end and end_commit.bellow != bellow:
        # Add CommitLink
        commit_link = to_commit(working_dir,
                                end_commit.bellow,
                                above_commit=end_commit,
                                level=commit.level + 1,
                                is_commit_link=True)
        result.append(commit_link)

    return result


def _to_rel_date(date) -> str:
    now = datetime.now(timezone.utc)
    delta = now - date
    _format = vcs.CONFIG['history']['author_date_format']
    try:
        return babel.dates.format_timedelta(delta, format=_format)
    except KeyError as exc:
        if delta.total_seconds() < 60:
            return f'{round(delta.total_seconds())} s'
        raise exc


def is_folded(commit_list: list[Commit], pos: int) -> bool:
    ''' Return true if the commit at pos is folded or not by looking at the next
        commit in the list.
    '''
    if pos >= len(commit_list):
        raise ValueError('Position %d should be less or equal the length %d' %
                         (pos, len(commit_list)))

    if pos == len(commit_list) - 1:
        return False
    assert commit_list[pos].is_merge  # nosec
    actual = commit_list[pos + 1].level
    expected = commit_list[pos].level
    return expected == actual


def _type_icon(bellow: Commit, is_commit_link: bool) -> str:
    ''' Return the graph icon '''
    if bellow is None:
        return "◉"
    if is_commit_link:
        return "⭞"
    return "●"


class CommitNotFound(Exception):
    ''' Thrown when following a link or searching fails '''


def find_non_link(working_dir: str, commit_list: list[Commit],
                  needle: Commit) -> int:
    '''
        Search recursively in a commit list for specified commit which is not a
        link. Will modify the commit_list if needed.
    '''
    needle_date = datetime.fromisoformat(needle.committer_date)
    for i, commit in enumerate(commit_list):
        if commit.is_commit_link:
            continue
        if commit.oid == needle.oid:
            return i
        commit_date = datetime.fromisoformat(commit.committer_date)
        if commit_date < needle_date:
            break
        if commit.is_merge and \
                merge_base(working_dir, needle.oid, commit.children[0]):
            # ↑ optimization for not descending just imported subtrees \
            children = child_history(working_dir, commit)
            try:
                tmp = find_non_link(working_dir, children, needle)
                commit_list[i + 1:i + 1] = children
                return tmp + i + 1
            except CommitNotFound:
                pass

    raise CommitNotFound


def follow(working_dir: str, commit_list: list[Commit], pos: int) -> int:
    '''
        Find link target in the commit_list. Will fill up commit list if
        needed.
    '''
    if pos > len(commit_list):
        raise ValueError('Position %d should be less then length %d' %
                         (pos, len(commit_list)))
    link: Commit = commit_list[pos]

    try:
        # May be the link is already in the commit list?
        return find_non_link(working_dir, commit_list, link)
    except CommitNotFound:
        pass

    last_commit = commit_list[-1]
    last_oid: str = last_commit.oid
    end = merge_base(working_dir, last_oid, link.oid)
    if not end:
        raise CommitNotFound('Commit has no mergebase')

    rev_range = "%s~1..%s~1" % (end, last_oid)
    tmp = commits_for_range(working_dir,
                            rev_range,
                            above_commit=last_commit,
                            level=commit_list[-1].level)
    if tmp:
        commit_list += tmp
        return find_non_link(working_dir, commit_list, link)

    raise CommitNotFound

File: src/commit.rs
// Copyright (C) 2021  Bahtiar `kalkin-` Gadimov <bahtiar@gadimov.de>
//
// This file is part of git-log-viewer
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

use crate::ui::base::search::Needle;
use url::Url;

use getset::Getters;
use git_wrapper::Repository;
use posix_errors::PosixError;
use std::ffi::OsStr;
use std::fmt::{Debug, Display, Formatter};
use std::path::PathBuf;

macro_rules! next_string {
    ($split:expr) => {
        $split.next().expect("Another split").to_owned()
    };
}

#[derive(Clone, Eq, PartialEq)]
pub struct Oid(pub String);

impl Display for Oid {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        f.write_str(&self.0)
    }
}

impl Debug for Oid {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let mut text = self.0.clone();
        text.truncate(8);
        f.write_str(&text)
    }
}

#[derive(Debug, Eq, PartialEq, Clone)]
pub struct GitRef(pub String);

impl Display for GitRef {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        f.write_str(&self.0)
    }
}

#[derive(Getters)]
pub struct Commit {
    #[getset(get = "pub")]
    id: Oid,
    #[getset(get = "pub")]
    short_id: String,
    #[getset(get = "pub")]
    author_name: String,
    #[getset(get = "pub")]
    author_email: String,
    #[getset(get = "pub")]
    author_date: String,
    #[getset(get = "pub")]
    author_rel_date: String,
    #[getset(get = "pub")]
    committer_name: String,
    #[getset(get = "pub")]
    committer_email: String,
    #[getset(get = "pub")]
    committer_date: String,
    #[getset(get = "pub")]
    committer_rel_date: String,
    #[getset(get = "pub")]
    subject: String,
    #[getset(get = "pub")]
    body: String,

    #[getset(get = "pub")]
    bellow: Option<Oid>,
    #[getset(get = "pub")]
    children: Vec<Oid>,
    #[allow(dead_code)]
    is_head: bool,
    #[allow(dead_code)]
    is_merge: bool,
    #[allow(dead_code)]
    branches: Vec<GitRef>,
    #[getset(get = "pub")]
    references: Vec<GitRef>,
    #[allow(dead_code)]
    tags: Vec<GitRef>,
}

impl Commit {
    #[must_use]
    pub fn is_merge(&self) -> bool {
        self.bellow.is_some() && !self.children.is_empty()
    }

    pub fn matches(&self, needle: &Needle) -> bool {
        let branches = &mut self.branches.iter().map(|v| &v.0).collect::<Vec<_>>();
        let tags = &mut self.tags.iter().map(|v| &v.0).collect::<Vec<_>>();
        let mut candidates = vec![
            self.author_name(),
            self.short_id(),
            &self.id().0,
            self.author_name(),
            self.author_email(),
            self.committer_name(),
            self.committer_email(),
            &self.subject,
        ];
        candidates.append(branches);
        candidates.append(tags);
        if *needle.ignore_case() {
            let needle_lowercase = needle.text();
            candidates
                .iter()
                .map(|x| x.to_lowercase())
                .any(|x| x.contains(needle_lowercase))
        } else {
            candidates.iter().any(|x| x.contains(needle.text()))
        }
    }
}

const REV_FORMAT: &str =
    "--format=%x1f%H%x1f%h%x1f%P%x1f%D%x1f%aN%x1f%aE%x1f%aI%x1f%ad%x1f%cN%x1f%cE%x1f%cI%x1f%cd%x1f%s%x1f%b%x1e";

impl Commit {
    #[must_use]
    pub fn new(data: &str) -> Self {
        let mut split = data.split('\x1f');
        split.next(); // skip commit: XXXX line
        let id = Oid(next_string!(split));

        let short_id = next_string!(split);
        let mut parents_record: Vec<&str> =
            split.next().expect("Parse parents").split(' ').collect();
        if parents_record.len() == 1 && parents_record[0].is_empty() {
            parents_record = vec![];
        }
        let references_record = next_string!(split);

        let author_name = next_string!(split);
        let author_email = next_string!(split);
        let author_date = next_string!(split);
        let author_rel_date = next_string!(split);

        let committer_name = next_string!(split);
        let committer_email = next_string!(split);
        let committer_date = next_string!(split);
        let committer_rel_date = next_string!(split);
        let subject = next_string!(split);
        let body = next_string!(split);

        let mut is_head = false;

        let mut references: Vec<GitRef> = Vec::new();
        let mut branches: Vec<GitRef> = Vec::new();
        let mut tags: Vec<GitRef> = Vec::new();
        for s in references_record.split(", ") {
            if s == "HEAD" {
                is_head = true;
            } else if s.starts_with("HEAD -> ") {
                is_head = true;
                let split_2: Vec<&str> = s.splitn(2, " -> ").collect();
                let branch = split_2[1].to_owned();
                branches.push(GitRef(branch.clone()));
                references.push(GitRef(branch));
            } else if s.starts_with("tag: ") {
                let split_2: Vec<&str> = s.splitn(2, ": ").collect();
                let tag = split_2[1].to_owned();
                tags.push(GitRef(tag.clone()));
                references.push(GitRef(tag));
            } else if s.is_empty() {
                // do nothing
            } else {
                let branch = s.to_owned();
                branches.push(GitRef(branch.clone()));
                references.push(GitRef(branch));
            }
        }

        let is_merge = parents_record.len() >= 2;

        let bellow = if parents_record.is_empty() {
            None
        } else {
            Some(Oid(parents_record.remove(0).to_owned()))
        };

        let mut children = Vec::new();
        for c in parents_record {
            children.push(Oid(c.to_owned()));
        }

        Self {
            id,
            short_id,
            author_name,
            author_email,
            author_date,
            author_rel_date,
            committer_name,
            committer_email,
            committer_date,
            committer_rel_date,
            subject,
            body,
            bellow,
            children,
            is_head,
            is_merge,
            branches,
            references,
            tags,
        }
    }

    pub fn parents(&self) -> &Vec<Oid> {
        self.children()
    }

    pub fn from_repo(repo: &Repository, oid: &Oid) -> Option<Self> {
        to_commit(repo, oid)
    }
}

/// Return commit count with `--first-parent`
///
/// # Errors
///
/// Returns a [`PosixError`] if `working_dir` does not exist or `rev_range` is invalid.
pub fn history_length<S>(
    repo: &Repository,
    rev_range: &Vec<S>,
    paths: &[PathBuf],
) -> Result<usize, PosixError>
where
    S: AsRef<OsStr>,
{
    let mut git = repo.git();
    git.args(vec!["rev-list", "--first-parent", "--count"])
        .args(rev_range)
        .arg("--")
        .args(paths);
    let proc = git.output().expect("Failed to run rev-list");

    if proc.status.success() {
        let text = String::from_utf8_lossy(&proc.stdout).trim_end().to_owned();
        return Ok(text
            .parse::<usize>()
            .expect("Failed to parse commit length"));
    }

    let err = PosixError::from(proc);
    if err.code() == 128 {
        let msg = "No revisions match the given arguments".to_owned();
        return Err(PosixError::new(128, msg));
    }
    Err(err)
}

#[allow(unused_qualifications)]
pub fn commits_for_range<S>(
    repo: &Repository,
    rev_range: &Vec<S>,
    paths: &[PathBuf],
    skip: Option<usize>,
    max: Option<usize>,
) -> Vec<Commit>
where
    S: AsRef<OsStr> + std::fmt::Debug,
{
    let mut cmd = repo.git();
    cmd.arg("rev-list")
        .args(vec!["--date=human", "--first-parent", REV_FORMAT]);

    let tmp;
    if let Some(val) = skip {
        tmp = format!("--skip={}", val);
        cmd.arg(&tmp);
    }

    let tmp2;
    if let Some(val) = max {
        tmp2 = format!("--max-count={}", val);
        cmd.arg(&tmp2);
    }

    cmd.args(rev_range);

    if !paths.is_empty() {
        cmd.arg("--");
        for p in paths {
            cmd.arg(<PathBuf as AsRef<OsStr>>::as_ref(p));
        }
    }

    let proc = cmd.output().expect("Failed to run git-rev-list(1)");
    if proc.status.success() {
        let output = String::from_utf8_lossy(&proc.stdout);
        let lines = output.split('\u{1e}');
        let mut result: Vec<Commit> = Vec::new();
        for data in lines {
            if data.is_empty() || data == "\n" {
                break;
            }
            result.push(Commit::new(data));
        }
        return result;
    }
    log::error!(
        "Failed to find commits for range({:?}), with skip({:?}) / max({:?}) & path({})",
        rev_range,
        skip,
        max,
        paths.is_empty()
    );
    vec![]
}

#[must_use]
pub fn child_history(repo: &Repository, commit: &Commit, paths: &[PathBuf]) -> Vec<Commit> {
    let bellow = commit.bellow.as_ref().expect("Expected merge commit");
    let first_child = commit.children.get(0).expect("Expected merge commit");
    let end = repo
        .merge_base(&[&bellow.0, &first_child.0])
        .expect("merge base shouldn't fail");

    let revision;
    if let Some(v) = &end {
        if v == &first_child.0 {
            revision = first_child.0.clone();
        } else {
            revision = format!("{}..{}", v, first_child.0);
        }
    } else {
        revision = first_child.0.clone();
    }
    commits_for_range(repo, &vec![revision], paths, None, None)
}

fn to_commit(repo: &Repository, oid: &Oid) -> Option<Commit> {
    let mut cmd = repo.git();
    cmd.args(["rev-list", "--date=human", REV_FORMAT, "-1", &oid.0]);
    let proc = cmd.output().expect("Failed to run git-rev-list(1)");
    proc.status.success().then(|| {
        let tmp = String::from_utf8_lossy(&proc.stdout);
        let lines: Vec<&str> = tmp.lines().collect();
        Commit::new(lines.get(1).expect("No second data line for commit"))
    })
}

pub fn parse_remote_url(input: &str) -> Option<Url> {
    // TODO handle upper case wording
    #[allow(clippy::case_sensitive_file_extension_comparisons)]
    let remote = if input.ends_with(".git") {
        input
            .strip_suffix(".git")
            .expect("Remote url ends with .git")
    } else {
        input
    };
    if let Ok(u) = Url::parse(remote) {
        return Some(u);
    }
    if remote.contains(':') {
        let tmp: Vec<&str> = remote.splitn(2, ':').collect();
        if tmp.len() == 2 {
            let domain = tmp[0];
            let path = tmp[1];
            let candidate = format!("ssh://{}/{}", domain, path);
            if let Ok(u) = Url::parse(&candidate) {
                return Some(u);
            }
        }
    }
    None
}

#[cfg(test)]
mod test {
    use pretty_assertions::assert_eq;
    use std::path::PathBuf;

    use crate::commit::commits_for_range;
    use git_wrapper::Repository;

    #[test]
    fn initial_commit() {
        let repo = Repository::default().unwrap();
        let paths: &[PathBuf] = &[];
        let result = commits_for_range(&repo, &vec!["a17989470af".to_owned()], paths, None, None);
        assert_eq!(result.len(), 1);
        let commit = &result[0];
        assert_eq!(commit.children.len(), 0);
        assert_eq!(commit.bellow, None);
        assert!(!commit.is_merge);
    }
}
